crates/agglayer-contracts/src/rollup.rs

```diff
@@ -64,22 +64,53 @@ where
     }
 
     async fn get_l1_info_root(&self, l1_leaf_count: u32) -> Result<[u8; 32], L1RpcError> {
+        // Check if we already have this l1_info_root cached
+        {
+            let cache = self.l1_info_roots.read().await;
+            if let Some(&cached_root) = cache.get(&l1_leaf_count) {
+                trace!(
+                    "Retrieved cached L1 info root for leaf count {}: {}",
+                    l1_leaf_count,
+                    alloy::primitives::B256::from(cached_root)
+                );
+                return Ok(cached_root);
+            }
+        }
+
         use alloy::sol_types::SolEvent;
 
         use crate::contracts::PolygonZkEvmGlobalExitRootV2::UpdateL1InfoTreeV2;
 
         // Get `UpdateL1InfoTreeV2` event for the given leaf count from the latest block
-        let filter = Filter::new()
-            .address(self.l1_info_tree)
-            .event_signature(UpdateL1InfoTreeV2::SIGNATURE_HASH)
-            .topic1(U256::from(l1_leaf_count))
-            .from_block(BlockNumberOrTag::Earliest);
-
-        let events = self
+        // To not hit the provider limit, we start from genesis and restrict search
+        // to the EVENT_FILTER_BLOCK_RANGE blocks range.
+        let mut events = Vec::new();
+        let mut start_block = 0u64;
```

Comment:
Same here

Reply:
> Same answer - we could as a followup. Please open the ticket. It would require binary search for when the address of the contract has code.

Yea, I don't know, binary search sounds rather annoying :smile: